Use Identifier instead of Strings

sapols opened the pull request:
I went through the repo and replaced every String that referred to an identifier with our real `Identifier`.

I'm pretty sure I caught everything. After making `MetadataLike.id` return `Identifier`, I was playing error whack-a-mole for a while. Then after I fixed those errors, I used file searches to search for any instance of the words `id`, `identifier`, `name`, or `vname` in the repo. Whenever I found something that was a String identifier, I replaced it. So, short of reading every line in every file myself, I can't find any more instances of String identifiers. 

Closes #156, #130, #128

sapols committed:
Refactor MetadataLike.id to return Option[Identifier]
and start fixing build errors

sapols committed:
Refactor MetadataLike.id to return Option[Identifier]
and start fixing build errors

sapols committed:
Fix remaining build errors that aren't tests

sapols committed:
Fix remaining build errors that aren't tests

sapols committed:
Add a TODO

sapols committed:
Add a TODO

sapols committed:
Style fix and clarify a TODO

sapols committed:
Style fix and clarify a TODO

sapols committed:
Style fixup

sapols committed:
Style fixup

sapols committed:
Remove TODO

sapols committed:
Remove TODO

sapols committed:
Add a TODO

sapols committed:
Add a TODO

sapols committed:
Replace comparison of Option and Some with .contains

sapols committed:
Replace comparison of Option and Some with .contains

sapols committed:
Replace pattern matches with folds

sapols committed:
Replace pattern matches with folds

sapols committed:
Replace another pattern match with a fold

sapols committed:
Replace another pattern match with a fold

sapols committed:
Consolidate if check with .exists

sapols committed:
Consolidate if check with .exists

sapols committed:
Fix Netcdf code errors

sapols committed:
Fix Netcdf code errors

sapols committed:
Fix test errors

sapols committed:
Fix test errors

sapols committed:
Fix compile error and add a TODO

sapols committed:
Fix compile error and add a TODO

sapols committed:
Another id fixup

sapols committed:
Another id fixup

sapols committed:
Allow dots in Identifiers

sapols committed:
Allow dots in Identifiers

sapols committed:
Fully refactor dataType with Identifiers

sapols committed:
Fully refactor dataType with Identifiers

sapols committed:
Don't capitalize "identifier" in exception messages

sapols committed:
Don't capitalize "identifier" in exception messages

sapols committed:
Start refactoring Operations

sapols committed:
Start refactoring Operations

sapols committed:
Move Identifier to new "macros" subproject

sapols committed:
Move Identifier to new "macros" subproject

sapols committed:
Continue using Identifier in new places

sapols committed:
Continue using Identifier in new places

sapols committed:
Use Identifier in even more new places

sapols committed:
Use Identifier in even more new places

sapols committed:
Refactor more operations

sapols committed:
Refactor more operations

sapols committed:
Fix a test

sapols committed:
Fix a test

sapols committed:
NetcdfAdapter refactoring

sapols committed:
NetcdfAdapter refactoring

sapols committed:
Fix another test

sapols committed:
Fix another test

sapols committed:
More test fixes

sapols committed:
More test fixes

sapols committed:
Fix more build/test errors

sapols committed:
Fix more build/test errors

sapols committed:
Replace NetcdfAdapter with master's copy

sapols committed:
Refactor NetcdfAdapter after rebase

sapols committed:
Merge remote-tracking branch 'origin/identifiers' into identifiers

# Conflicts:
#	netcdf/src/main/scala/latis/input/NetcdfAdapter.scala

sapols committed:
Refactor DatasetResolver to use Identifier

sapols committed:
Correct parser/dap2 errors

sapols committed:
Test travis build without ConstraintParserProps

sapols committed:
Use Identifier with FDML

sapols committed:
Go all in with Identifier in NetcdfAdapter

sapols committed:
Manually pull in changes from master and fix errors

sapols committed:
Fix copy/paste indentation

sapols committed:
Merge branch 'master' into identifiers

sapols commented on dap2-service/src/test/scala/latis/util/dap2/parser/ConstraintParserProps.scala:
```diff
@@ -1,104 +1,111 @@
+//
+//  val cexpr: Gen[ConstraintExpression] =
+//    Gen.listOf(expr).map(ConstraintExpression)
+//}
```
This `ConstraintParserProps` is the last file causing me grief. I commented it out just to prove to myself that the Travis build succeeds without it. 

It's hard to nail down what's wrong because the random nature of this thing means the error message changes each time I run it. I'll paste one of the shorter messages I've gotten:

> failing seed for DAP 2 Constraint Parser.parse is WVDGv1BsR1YBfraMGSE9clHpcLfuNYTE0dIftPt_czP=
> ! DAP 2 Constraint Parser.parse: Falsified after 2 passed tests.
> > ARG_0: ConstraintExpression(List(Projection(List())))
> > ARG_0_ORIGINAL: ConstraintExpression(List(Projection(List(Identifier(vq_)
>   )), Selection(Identifier(l),Eq,"xc")))
> Found 1 failing properties.
>  

I don't even know how to interpret that message. I can tell I broke it though because I usually see "Identifier" somewhere in the errors. Lol

lindholc replied:
It looks like there are places where an `Identifier` is being turned into a `String` automatically with `toString` rather than `asString`:

- in the `variable` generator
- in the `pretty` method

sapols replied:
Eureka! Thank you.

lindholc commented on build.sbt:
```diff
@@ -97,6 +98,7 @@ lazy val latis = project
   )
 
 lazy val core = project
```
I think you can remove the dependency on `scala-reflect` from `core`.

sapols replied:
Yup. 👍

sapols committed:
Fix ConstraintParserProps using Identifier.asString

sapols committed:
Merge remote-tracking branch 'origin/identifiers' into identifiers

dlindhol commented on core/src/main/scala/latis/dataset/package.scala:
```diff
@@ -15,12 +16,12 @@ package object dataset {
     def compose(df: Dataset): Dataset = dataset.withOperation(Composition(df))
-    def contains(varName: String, values: String*): Dataset = dataset.withOperation(Contains(varName, values: _*))
-    def rename(varName: String, newName: String): Dataset = dataset.withOperation(Rename(varName, newName))
+    def contains(varName: Identifier, values: String*): Dataset = dataset.withOperation(Contains(varName, values: _*))
```
Should we change the variable names? `varId`? `id`?

sapols replied:
I'm down to change them. `varId` here sounds good, and makes it clearer that they're identifiers.

sapols replied:
We decided to go with `id` everywhere.

dlindhol commented on core/src/main/scala/latis/model/dataType.scala:
```diff
@@ -64,9 +65,9 @@ sealed trait DataType extends MetadataLike with Serializable {
-        findAllVariables(n1).flatMap(_.findAllVariables(n2.mkString(".")))
-      }
+      case Array(n1, n2 @ _*) =>
+        findAllVariables(Identifier.fromString(n1).get)
```
`get` smells. Maybe add a comment that it is safe since `n1` came from a valid `Identifier`?

sapols replied:
Good call.

dlindhol commented on core/src/main/scala/latis/ops/Projection.scala:
```diff
@@ -1,14 +1,17 @@
  * Domain variables will be included, for now.
  */
-case class Projection(vnames: String*) extends MapOperation {
+case class Projection(vnames: Identifier*) extends MapOperation {
```
OK, so far I see that, in the past, we have used `vname`, `varName`, and `variableName`. Relating to an earlier comment, should we just use `id` everywhere?

sapols replied:
Yes. 👍

dlindhol commented on core/src/main/scala/latis/ops/Projection.scala:
```diff
@@ -56,11 +59,21 @@ case class Projection(vnames: String*) extends MapOperation {
-  def apply(exp: String): Projection =
-    Projection(exp.split(",").toIndexedSeq: _*)
+  def apply(exp: String): Projection = {
+    val ids = exp.split(",").map(id => Identifier.fromString(id).getOrElse {
```
Whenever I see `map` with parens on different lines I'd be inclined to use "{}" with a new line after the opening "{".

It was the "})" below that stood out.

sapols replied:
I'll change this.

dlindhol commented on core/src/main/scala/latis/ops/Projection.scala:
```diff
@@ -56,11 +59,21 @@ case class Projection(vnames: String*) extends MapOperation {
+      case Nil => throw LatisException("Projection requires at least one argument")
+      case _ => Projection(
+        args.map(id =>
+          Identifier.fromString(id).getOrElse(throw LatisException(s"'$id' is not a valid identifier"))
```
Throwing an exception and using `Either.catchOnly` feels like cheating.

lindholc replied:
Yeah, I feel like it would be better to just wrap values in `Right` or `Left`.

sapols replied:
While I share this sentiment, I couldn't figure out a clean way to code this without `Either.catchOnly`. I couldn't just wrap values in `Right` or `Left` because of the placement of the `getOrElse` on this line. 

I'm mapping over `args` to try to turn the Strings into Identifiers. If any of them fail, we should get an exception instead of a `Projection`. That's the behavior here. If I just wrap the exception in `Left`, `args` turns into something that contains either `Right(Identifier)`s or `Left(LatisException)`s, which is nonsensical and that short-circuiting exception behavior is gone.

lindholc replied:
I think you'd want something like:

```scala
args.traverse { id =>
  Identifier.fromString(id).toRight(LatisException(...))
}.map(Projection(_:_*))
```

sapols replied:
Done.

dlindhol commented on core/src/main/scala/latis/ops/Rename.scala:
```diff
@@ -32,8 +34,14 @@ case class Rename(origName: String, newName: String) extends UnaryOperation {
+  def fromArgs(args: List[String]): Either[LatisException, Rename] = Either.catchOnly[LatisException] {
+    args.map(arg => Identifier.fromString(arg)
+      .getOrElse {
+        throw LatisException(s"'$arg' is not a valid identifier")
```
Another one.

dlindhol commented on core/src/main/scala/latis/ops/Selection.scala:
```diff
@@ -96,12 +97,16 @@ object Selection {
   def apply(expression: String): Selection = {
```
Looks like we need safer construction that returns `Either`.
#161 should address this.

dlindhol commented on core/src/main/scala/latis/util/CacheManager.scala:
```diff
@@ -10,29 +10,32 @@ import latis.dataset.{Dataset, MemoizedDataset}
    * Singleton instance of the CacheManager.
    */
-  private lazy val cache = mutable.Map[String, Dataset]()
+  private lazy val cache = mutable.Map[Identifier, Dataset]()
```
I assume we are getting appropriate Map key support (e.g. equals and hashCode) via `Identifier` being a case class?

sapols replied:
I assume so. I did independent tests to confirm that Identifiers made from equal Strings do in fact equal each other.

dlindhol replied:
`Map`s often use the `hashCode` to determine which bucket things live in so it is important that it is consistent with the behavior of equals. Based on a little googling, case classes do provide consistent `equals` and `hashCode`.

dlindhol commented on core/src/test/scala/latis/dataset/DatasetSpec.scala:
```diff
@@ -8,12 +8,13 @@ import latis.data.Data._
 import latis.metadata.Metadata
 import latis.model._
 import latis.ops.Selection
+import latis.util.Identifier.IdentifierStringContext
```
Any thoughts on how we could provide a more convenient import mechanism?

lindholc replied:
It's only marginally better, but this works:

```scala
import latis.util.Identifier._
```

In the future, we could provide an `implicits` package, so someone could import `latis.implicits._` to get things like the `id` string context and the dataset ops extension methods.

sapols committed:
Comment safe usage of .get

sapols committed:
Use curlies instead of parens for .maps

sapols committed:
Remove scala-reflect dependency from core

sapols committed:
Standardize on "id" for identifier variable/param names

sapols commented:
Just standardized on `id` for identifier variable/parameter names. With that, I've addressed all comments and I'm done making changes. Review away. 👍
